### PR TITLE
#10843 ESR Import Error for new QR Invoice 

### PR DESCRIPTION
**Pull request by cristinamghita**
#10843 Implement QRR import support

### CODE REVIEW

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRType.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	{
+		ESRType type = typesByCode.get(code);
+		if (type == null)
+		{
+			throw new AdempiereException("No " + ESRType.class + " found for code: " + code);
+		}
+		return type;
+	}
+	
+	private static final ImmutableMap<String, ESRType> typesByCode = Maps.uniqueIndex(Arrays.asList(values()), ESRType::getCode);
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRType.java:60]: pls use ReferenceListAwareEnums.index() instead

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRDataLoaderUtil.java [hunk 2]
```diff
@@ -415,27 +416,53 @@ public void evaluateESRAccountNumber(final I_ESR_Import esrImport, final I_ESR_I
...
+		
+		
+		if (isQRR(importLine))
+		{
+			final String QR_IBAN = StringUtils.trimBlankToNull(bankAccount.getQR_IBAN());
+			final String IBAN = StringUtils.trimBlankToNull(bankAccount.getIBAN());
+			final String SEPA_CreditorIdentifier = StringUtils.trimBlankToNull(bankAccount.getSEPA_CreditorIdentifier());
+			
+			final String postAcctNoCleaned = StringUtils.cleanWhitespace(postAcctNo);
+			if (! (postAcctNoCleaned.equals(QR_IBAN)
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRDataLoaderUtil.java:428]: why not having a method BankAccount.isAccountNoMatching(String accountNo) ?

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRDataLoaderUtil.java [hunk 3]
```diff
@@ -415,27 +416,53 @@ public void evaluateESRAccountNumber(final I_ESR_Import esrImport, final I_ESR_I
...
-			ESRDataLoaderUtil.addMatchErrorMsg(importLine, Services.get(IMsgBL.class).getMsg(Env.getCtx(), ERR_WRONG_POST_BANK_ACCOUNT,
-					new Object[] { postAcctNo }));
+				ESRDataLoaderUtil.addMatchErrorMsg(importLine, Services.get(IMsgBL.class).getMsg(Env.getCtx(), ERR_WRONG_POST_BANK_ACCOUNT,
+						new Object[] { postAcctNo }));
+			}
 		}
 
 	}
+	
+	private boolean isQRR(@NonNull final I_ESR_ImportLine importLine)
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRDataLoaderUtil.java:461]: static

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRTransaction.java [hunk 4]
```diff
@@ -67,6 +67,9 @@
 	@NonNull
 	private final String trxType;
 	
+	@NonNull
+	private final String type;
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRTransaction.java:71]: why not using `ESRType`?

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java [hunk 5]
```diff
@@ -198,7 +232,37 @@ public void extractAndSetEsrReference(
 		return esrReferenceNumberString;
 	}
 	
+	private Optional<String> extractType(@NonNull final EntryTransaction8 txDtls)
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java:235]: return Optional<ESRType>

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java [hunk 6]
```diff
@@ -198,7 +232,37 @@ public void extractAndSetEsrReference(
...
+						&& cdtrRefInf.getTp() != null
+						&& cdtrRefInf.getTp().getCdOrPrtry() != null
+						&& (cdtrRefInf.getTp().getCdOrPrtry().getPrtry().equals(ESRType.TYPE_ESR.getCode())
+								|| cdtrRefInf.getTp().getCdOrPrtry().getPrtry().equals(ESRType.TYPE_QRR.getCode())))
+
+				.map(cdtrRefInf -> cdtrRefInf.getTp().getCdOrPrtry().getPrtry())
+				.findFirst();
+	}
 
+	private Optional<String> extractType(@NonNull final EntryTransaction2 txDtls)
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java:250]: return Optional<ESRType>

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java [hunk 7]
```diff
@@ -138,11 +170,12 @@ public void extractAndSetEsrReference(
 		final Optional<String> esrReferenceNumberString = txDtls.getRmtInf().getStrd().stream()
 				.map(strd -> strd.getCdtrRefInf())
 
-				// it's stored in the cdtrRefInf records whose cdtrRefInf/tp/cdOrPrtry/prtr equals to ISR_REFERENCE
-				.filter(cdtrRefInf -> cdtrRefInf != null
-						&& cdtrRefInf.getTp() != null
+				// it's stored in the cdtrRefInf records whose cdtrRefInf/tp/cdOrPrtry/prtr equals to ISR_REFERENCE or QRR_REFERENCE
+				.filter(cdtrRefInf -> cdtrRefInf != null
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java:174]: i guess u can extract in a method the whole expression from filter(...).
Saw it that u are using it in couple of places.

File: backend/de.metas.payment.esr/src/test/java/de/metas/payment/esr/dataimporter/impl/camt54/ESRDataImporterCamt54V02Tests.java [hunk 8]
```diff
@@ -2,21 +2,24 @@
...
 import java.io.InputStream;
 import java.math.BigDecimal;
 
+import org.adempiere.exceptions.AdempiereException;
 import org.adempiere.test.AdempiereTestHelper;
 import org.assertj.core.api.Condition;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
+import org.junit.Before;
+import org.junit.Test;
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/test/java/de/metas/payment/esr/dataimporter/impl/camt54/ESRDataImporterCamt54V02Tests.java:14]: why converting from junit5 to junit4?

Comment (cristinamghita) [reply backend/de.metas.payment.esr/src/test/java/de/metas/payment/esr/dataimporter/impl/camt54/ESRDataImporterCamt54V02Tests.java:14]: by misstake

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRDataLoaderUtil.java [hunk 9]
```diff
@@ -415,27 +416,43 @@ public void evaluateESRAccountNumber(final I_ESR_Import esrImport, final I_ESR_I
 		final BankAccount bankAccount = bpBankAccountRepo.getById(BankAccountId.ofRepoId(bankAccountRecordId));
 
 		final String postAcctNo = importLine.getESRPostParticipantNumber();
+		
+		if (isQRR(importLine) && !bankAccount.isAccountNoMatching(postAcctNo))
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/ESRDataLoaderUtil.java:420]: i think this logic is not corrent

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java [hunk 10]
```diff
@@ -167,17 +199,17 @@ public void extractAndSetEsrReference(
...
 				.map(strd -> strd.getCdtrRefInf())
 
-				// it's stored in the cdtrRefInf records whose cdtrRefInf/tp/cdOrPrtry/prtr equals to ISR_REFERENCE
+				// it's stored in the cdtrRefInf records whose cdtrRefInf/tp/cdOrPrtry/prtr equals to ISR_REFERENCE or QRR_REFERENCE
 				.filter(cdtrRefInf -> cdtrRefInf != null
 						&& cdtrRefInf.getTp() != null
 						&& cdtrRefInf.getTp().getCdOrPrtry() != null
-						&& cdtrRefInf.getTp().getCdOrPrtry().getPrtry().equals(ISR_REFERENCE))
-
+						&& isSupportedESRType(cdtrRefInf))
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java:206]: same here

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java [hunk 11]
```diff
@@ -167,17 +199,17 @@ public void extractAndSetEsrReference(
...
+				.filter(cdtrRefInf -> cdtrRefInf != null)
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java:207]: i think this is redundant

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java [hunk 12]
```diff
@@ -197,8 +229,52 @@ public void extractAndSetEsrReference(
...
 		return esrReferenceNumberString;
 	}
-	
 
+	private Optional<ESRType> extractType(@NonNull final EntryTransaction8 txDtls)
+	{
+		return txDtls.getRmtInf().getStrd().stream()
+				.map(strd -> strd.getCdtrRefInf())
+
+				.filter(cdtrRefInf -> cdtrRefInf != null
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java:238]: use your new method

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java [hunk 13]
```diff
@@ -197,8 +229,52 @@ public void extractAndSetEsrReference(
...
+				.findFirst();
+
+	}
+
+	private Optional<ESRType> extractType(@NonNull final EntryTransaction2 txDtls)
+	{
+		return txDtls.getRmtInf().getStrd().stream()
+				.map(strd -> strd.getCdtrRefInf())
+
+				.filter(cdtrRefInf -> cdtrRefInf != null
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java:254]: us your new method

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java [hunk 14]
```diff
@@ -197,8 +229,52 @@ public void extractAndSetEsrReference(
...
+						&& cdtrRefInf.getTp().getCdOrPrtry() != null
+						&& isSupportedESRType(cdtrRefInf))
+
+				.map(cdtrRefInf -> cdtrRefInf.getTp().getCdOrPrtry().getPrtry())
+				.map(r ->  ESRType.ofNullableCode(r))
+				.findFirst();
+
+	}
+
+	private static boolean isSupportedESRType(CreditorReferenceInformation2 cdtrRefInf)
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ReferenceStringHelper.java:265]: better use FQ class name

**Pull request merged**